src/main/java/frc/robot/systems/IntakeFSMSystem.java

```diff
@@ -0,0 +1,707 @@
+					if (input.isIntakeButtonPressed() && !input.isOuttakeButtonPressed()
+						&& !input.isShootButtonPressed()) {
+						return IntakeFSMState.INTAKING;
+					} else if (input.isOuttakeButtonPressed() && !input.isIntakeButtonPressed()
+								&& !input.isShootButtonPressed()) {
+						return IntakeFSMState.OUTTAKING;
+					}
+				}
+
+				if ((input.isIntakeButtonPressed() || input.isOuttakeButtonPressed())
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				} else {
+					return IntakeFSMState.MOVE_TO_HOME;
+				}
+
+			case INTAKING:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)
+					&& input.isIntakeButtonPressed() && !input.isOuttakeButtonPressed()
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.INTAKING;
+				}
+
+				if (input.isShootButtonPressed()
+					&& !(input.isOuttakeButtonPressed() || input.isIntakeButtonPressed())) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			case OUTTAKING:
+				if (inRange(throughBore.getDistance(), Constants.GROUND_ENCODER_COUNT)
+					&& input.isOuttakeButtonPressed() && !input.isIntakeButtonPressed()
+					&& !input.isShootButtonPressed()) {
+					return IntakeFSMState.OUTTAKING;
+				}
+
+				if (input.isShootButtonPressed()
+					&& !(input.isOuttakeButtonPressed() || input.isIntakeButtonPressed())) {
+					return IntakeFSMState.MOVE_TO_HOME;
+				} else {
+					return IntakeFSMState.MOVE_TO_GROUND;
+				}
+
+			default:
+				throw new IllegalStateException("Invalid state: " + currentState.toString());
+		}
+	}
+
+/* ------------------------ Command handlers ------------------------ */
+
+	private boolean inRange(double targetEncoder, double currentEncoder) {
+		return Math.abs(targetEncoder - currentEncoder) < Constants.INRANGE_VALUE;
+	}
+
+	/**
+	 * A PID controller for the pivot motor.
+	 * @param currentEncoderPID The current position of the pivot motor in encoder units.
+	 * @param targetEncoder The target position of the pivot motor in encoder units.
+	 * @return The correct power to send to the pivot motor to achieve the target position.
+	 */
+	private double pid(double currentEncoderPID, double targetEncoder) {
+		double correction = Constants.PID_CONSTANT_PIVOT_P
+			* (targetEncoder - currentEncoderPID);
+		return Math.min(Constants.MAX_TURN_SPEED,
+			Math.max(Constants.MIN_TURN_SPEED, correction));
+	}
+
+	/**
+	 * A PID controller for the pivot motor in autonomous mode.
+	 * @param currentEncoderPID The current position of the pivot motor in encoder units.
+	 * @param targetEncoder The target position of the pivot motor in encoder units.
+	 * @return The correct power to send to the pivot motor to achieve the target position.
+	 */
+	private double pidAuto(double currentEncoderPID, double targetEncoder) {
+		double correction = Constants.PID_CONSTANT_PIVOT_P_AUTO
```

Comment:
WPILib also provides a PIDController class if you don't want to write up the math yourself for I and D. Although if P only is working for you, then no need to overcomplicate things. https://docs.wpilib.org/en/stable/docs/software/advanced-controls/controllers/pidcontroller.html

Reply:
I believe we tried using the WPILib PIDController at one point, but we were having issues tuning it. We also don't really need to use the I or D so a P only controller works just fine. We could try using the CTRE MotionMagic controller after ccc as a project since we now have access to krakens.